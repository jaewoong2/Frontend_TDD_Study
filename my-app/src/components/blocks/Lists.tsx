import { Button } from 'components/atoms';
import { Label } from 'components/atoms/Label';
import React from 'react';
import styled from 'styled-components';

type Props = {
  lists: string[];
  handleDelete: (index: number) => void;
};

function Lists({ lists, handleDelete }: Props) {
  return (
    <Container>
      {lists.map((list, i) => (
        <ItemContainer key={`${list + i}`}>
          <Label>{list ?? 'hi'}</Label>
          <Button
            onClick={() => handleDelete(i)}
            hoverColor="#F01440"
            backgroundColor="#FF1744">
            Delete
          </Button>
        </ItemContainer>
      ))}
    </Container>
  );
}

export default Lists;

const Container = styled.ul`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
`;

const ItemContainer = styled.li`
  display: flex;
  align-items: center;
  border-bottom: 1px solid #bdbdbd;
  margin: 10px;
  padding: 10px;

  label {
    flex: 1;
    color: black;
  }
`;
