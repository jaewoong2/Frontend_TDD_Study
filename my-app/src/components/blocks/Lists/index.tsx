import { Item } from 'components/atoms';
import React, { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import styled from 'styled-components';

type Props = {
  lists: string[];
};

function Lists({ lists }: Props) {
  const navigate = useNavigate();

  const handleClickItem = useCallback((index: number) => {
    navigate(`/detail/${index}`);
  }, []);

  return (
    <Container data-testid="toDoList">
      {lists.map((list, i) => (
        <Item
          key={`${list + i}`}
          backgroundColor="#d9d9d9"
          hoverColor="#d9d9d9"
          onClick={() => handleClickItem(i)}
          buttonName="Move to Item">
          {list}
        </Item>
      ))}
    </Container>
  );
}

export default Lists;

const Container = styled.ul`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
`;
