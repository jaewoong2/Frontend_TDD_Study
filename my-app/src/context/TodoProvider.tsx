import React, { createContext, useEffect, useMemo, useState } from 'react';

const initalAction = {
  addList: () => (todo: string) => {},
  removeList: (index: number) => {},
};

export const TodoActionsContext = createContext(initalAction);
export const TodoValueContext = createContext<string[]>([]);

type Props = {
  children?: React.ReactNode;
};

export function TodoProvider({ children }: Props) {
  const { addStroage, getStorage, removeStroage } = utils;
  const [todoLists, setTodoLists] = useState<string[]>(getStorage());

  const actions = useMemo(
    () => ({
      addList: () => (todo: string) => {
        if (todo.trim() !== '') {
          setTodoLists((prev) => [...prev, todo]);
          addStroage(todo);
        }
      },
      removeList: (index: number) =>
        setTodoLists((prev) => {
          removeStroage(index);
          return [...prev].filter((_, i) => i !== index);
        }),
    }),
    [],
  );

  return (
    <TodoActionsContext.Provider value={actions}>
      <TodoValueContext.Provider value={todoLists}>
        {children}
      </TodoValueContext.Provider>
    </TodoActionsContext.Provider>
  );
}

const utils = {
  getStorage: () => {
    const data = localStorage.getItem('todoItem');
    if (data !== null) {
      const todolists = JSON.parse(data);
      if (todolists) {
        return todolists;
      }
      return [];
    }
    return [];
  },

  addStroage: (item: string) => {
    const data = utils.getStorage();
    const lists = [...data, item];
    localStorage.setItem('todoItem', JSON.stringify(lists));
  },

  removeStroage: (index: number) => {
    const data = utils.getStorage();
    const lists = [...data].filter((_, i) => i !== index);
    localStorage.setItem('todoItem', JSON.stringify(lists));
  },
};
