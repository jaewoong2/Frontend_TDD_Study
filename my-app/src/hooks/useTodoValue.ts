import { useContext } from 'react';
import { TodoValueContext } from 'context/TodoProvider';

export default () => {
  const values = useContext(TodoValueContext);
  if (values === undefined) {
    throw new Error('useCounterValue should be used within CounterProvider');
  }
  return values;
};
