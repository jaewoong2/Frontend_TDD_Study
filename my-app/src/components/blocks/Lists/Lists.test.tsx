import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { TodoProvider } from 'context/TodoProvider';
import useTodoValue from 'hooks/useTodoValue';
import { BrowserRouter } from 'react-router-dom';
import Lists from './index';

beforeEach(() => {
  localStorage.clear();
});

describe('<Lists />', () => {
  it('render List correctly', () => {
    const { container } = renderWithRouter(<Lists lists={['default value']} />);

    const todoItem = screen.getByText('default value');
    const button = screen.getByText('Move to Item');

    expect(todoItem).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('click the button to detail', () => {
    renderWithRouter(<Lists lists={['default value']} />);

    const todoItem = screen.getByText('default value');
    const button = screen.getByText('Move to Item');

    expect(todoItem).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    fireEvent.click(button);
    expect(window.location.pathname).toBe('/detail/0');
  });
});

const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
