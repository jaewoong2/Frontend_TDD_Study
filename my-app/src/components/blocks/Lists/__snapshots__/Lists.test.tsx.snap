// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lists /> render List correctly 1`] = `
<div>
  <ul
    class="sc-gKXOVf eJVsLV"
    data-testid="toDoList"
  >
    <li
      class="sc-eCYdqJ ciJnwK"
    >
      <label
        class="sc-bczRLJ LpNGc"
      >
        default value
      </label>
      <button
        class="sc-gsnTZi bWpsoW"
      >
        <label
          class="sc-bczRLJ LpNGc"
        >
          Move to Item
        </label>
      </button>
    </li>
  </ul>
</div>
`;
